lib/rucio/tests/test_dataset_replicas.py

```diff
@@ -124,3 +127,211 @@ def test_list_dataset_replicas_archive(self):
+        update_collection_replica(update_request_id=update_request.id, session=self.db_session)
+        update_request = self.db_session.query(models.UpdatedCollectionReplica).filter_by(id=update_request.id).first()
+        assert_equal(update_request, None)
+        dataset_replica = self.db_session.query(models.CollectionReplica).filter_by(scope=self.scope, name=dataset_name).one()
+        assert_equal(dataset_replica['bytes'], len(files) * file_size)
+        assert_equal(dataset_replica['length'], len(files))
+        assert_equal(dataset_replica['available_bytes'], len(files) * file_size)
+        assert_equal(dataset_replica['available_replicas_cnt'], len(files))
+        assert_equal(str(dataset_replica['state']), 'AVAILABLE')
+
+        # Delete one file replica -> dataset replica should be unavailable
+        delete_replicas(rse=self.rse, files=[files[0]], session=self.db_session)
+        update_request = self.db_session.query(models.UpdatedCollectionReplica).filter_by(rse_id=self.rse_id, scope=self.scope, name=dataset_name).one()
+        update_collection_replica(update_request_id=update_request.id, session=self.db_session)
+        dataset_replica = self.db_session.query(models.CollectionReplica).filter_by(scope=self.scope, name=dataset_name).one()
+        assert_equal(dataset_replica['bytes'], len(files) * file_size)
+        assert_equal(dataset_replica['length'], len(files))
+        assert_equal(dataset_replica['available_bytes'], (len(files) - 1) * file_size)
+        assert_equal(dataset_replica['available_replicas_cnt'], len(files) - 1)
+        assert_equal(str(dataset_replica['state']), 'UNAVAILABLE')
+
+        # Add one file replica -> dataset replica should be available again
+        add_replicas(rse=self.rse, files=[files[0]], account='root', session=self.db_session)
+        attach_dids(scope=self.scope, name=dataset_name, dids=[files[0]], account='root', session=self.db_session)
+        models.UpdatedCollectionReplica(rse_id=self.rse_id, scope=self.scope, name=dataset_name, did_type=constants.DIDType.DATASET).save(session=self.db_session)
+        update_request = self.db_session.query(models.UpdatedCollectionReplica).filter_by(rse_id=self.rse_id, scope=self.scope, name=dataset_name).one()
+        update_collection_replica(update_request_id=update_request.id, session=self.db_session)
+        dataset_replica = self.db_session.query(models.CollectionReplica).filter_by(scope=self.scope, name=dataset_name).one()
+        assert_equal(dataset_replica['bytes'], len(files) * file_size)
+        assert_equal(dataset_replica['length'], len(files))
+        assert_equal(dataset_replica['available_bytes'], len(files) * file_size)
+        assert_equal(dataset_replica['available_replicas_cnt'], len(files))
+        assert_equal(str(dataset_replica['state']), 'AVAILABLE')
+
+        # Delete all file replicas -> dataset replica should be deleted
```

Comment:
There is a difference between replica updates with and with out rse ids. If all file replicas of a collection are deleted, a collection replica gets also deleted if the replica update has a rse id, which makes sense to me. But the collection replica is not deleted if the replica update has no rse id. In this case nothing is happening, according to the procedure?

Reply:
Yes, this is a bit specific to the workflow how the UpdatedCollectionReplicas are created. Essentially UpdatedCollectionReplicas with the rse_id not set, are never created when replicas are reduced, but only when they are increased. So the case where the CollectionReplica becomes size=0 is only when the rse_id of the UpdatedCollectionReplica is set.